Provide a method for generating variations based off of product attributes.

justinshreve opened the pull request:
This PR introduces a helper method for generating variation objects based on a product's available attributes.

It accepts a product object and will return an array of API-ready variation objects that can later be used for actually creating the variations in state to allow editing of details.

It includes tests and an example.

I am putting it in a `lib/` directory where these types of methods can live (per https://github.com/Automattic/wp-calypso/pull/12702#issuecomment-292266600).

To Test:
* Run `make test` and make sure all tests pass.

justinshreve commented on client/extensions/woocommerce/lib/generate-variations/index.js:
```diff
@@ -0,0 +1,33 @@
+const f = ( a, b ) => [].concat( ...a.map( c => b.map( d => [].concat( c, d ) ) ) );
+const cartesian = ( a, b, ...c ) => b ? cartesian( f( a, b ), ...c ) : a;
+
+export default function generateVariations( product ) {
```
Just a note to add a doc here when I am on later this morning.

psealock commented:
I've looked through this code and it all LGTM. 

Just a nitpick here: I see you filter for `attribute.variation`. One more test might be useful to future proof against changes to that behavior, ie `attribute.variation` is set to false.

justinshreve commented:
Thanks for the review @psealock. I have added an additional test that passes a non-variation attribute.

coderkevin commented on client/extensions/woocommerce/lib/generate-variations/index.js:
```diff
@@ -0,0 +1,39 @@
+	const variationTypes = ( attributes && attributes.filter( attribute => attribute.variation ) ) || [];
+
+	if ( variationTypes.length < 2 ) {
+		return variationTypes[ 0 ] && variationTypes[ 0 ].options.map( function( option ) {
```
If `variationTypes.length === 0` this will throw an exception.

coderkevin commented on client/extensions/woocommerce/lib/generate-variations/index.js:
```diff
@@ -0,0 +1,39 @@
+	const { attributes } = product;
+	const variationTypes = ( attributes && attributes.filter( attribute => attribute.variation ) ) || [];
+
+	if ( variationTypes.length < 2 ) {
```
Why the special handling for a single variation type?

justinshreve replied:
1) I thought we could just short circuit because if there is a single type, the variations are just that type. 2) I thought the API response for variations with one single attribute was slightly different. 

I just went back and checked number 2 and that turned out to be wrong so this doesn't seem as necessary.

coderkevin commented on client/extensions/woocommerce/lib/generate-variations/test/index.js:
```diff
@@ -0,0 +1,122 @@
+ */
+import generateVariations from '../index';
+
+describe( 'generateVariations', () => {
```
Maybe add a test with 3+ variation types to make sure the code holds up?

justinshreve replied:
@coderkevin https://github.com/Automattic/wp-calypso/pull/13436/files#diff-c35d147d05f7b3cae0bacb4753172e53R96 does that. It has three types: Color, Size, Cut and generates 4 combos:

Red - Small - Unisex
Red - Medium - Unisex
Blue - Small - Unisex
Blue - Medium - Unisex

Do you think I need to expand this at all?

coderkevin replied:
I guess I meant to have 2+ options on 3+ types.

justinshreve replied:
Okay I've updated that test to add an additional color, and an additional option on the third type -- giving us 12 variations built on 3 options: Color - 3 Types, Size - 2 Types, Sleeves - 2 Types.

coderkevin replied:
Works for me, thanks!

justinshreve commented:
@coderkevin Thanks for the review so far. I've made a few changes here and left a comment about the 3+ test.

coderkevin commented on client/extensions/woocommerce/lib/generate-variations/index.js:
```diff
@@ -0,0 +1,39 @@
+		attributes.filter( attribute => attribute.variation && attribute.name && attribute.options.length > 0 )
+	) || [];
+
+	if ( variationTypes.length < 1 ) {
```
I don't think this is necessary. An empty array will just fall out of the `forEach()` loop and `cartesian().map()` loop quickly.

justinshreve replied:
`cartesian()` currently returns undefined if you pass it nothing to work with -- however I've updated the return statement to handle this differently.

coderkevin replied:
👍

coderkevin commented:
Looks good, thanks Justin! :shipit:

DanReyLop commented on client/extensions/woocommerce/lib/generate-variations/index.js:
```diff
@@ -0,0 +1,36 @@
+// http://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript
+const f = ( a, b ) => [].concat( ...a.map( c => b.map( d => [].concat( c, d ) ) ) );
+const cartesian = ( a, b, ...c ) => b ? cartesian( f( a, b ), ...c ) : a;
```
I don't know if this will become a performance concern (you talked in Slack that this may be executed on every keypress when editing a product attrs). There are clearly ways to optimize.

This solution creates *a lot* of temporary arrays, which will stress the GC. It may be worth to change it to a solution without so many `.map()` and `.concat()`, like http://stackoverflow.com/a/29585704 or https://pastebin.com/YbhqZuf7

DanReyLop replied:
Also, it probably will be more performant to construct first the `{ name:"Size", option:"Small" }` objects, then do the `cartesian product` over those whole objects. The objects won't be copied, so the RAM footprint of the (possibly massive) final array will be smaller. Plus, you won't need to iterate over the array again to add the variation names.

DanReyLop commented on client/extensions/woocommerce/lib/generate-variations/index.js:
```diff
@@ -0,0 +1,36 @@
+ * @return {Array} Array of variation objects.
+ */
+export default function generateVariations( product ) {
+	const { attributes } = product;
```
You could write `export default function generateVariations( { attributes } ) {` instead, I personally find that syntax one of the best shortcuts of ES6 :)

DanReyLop commented on client/extensions/woocommerce/lib/generate-variations/index.js:
```diff
@@ -0,0 +1,36 @@
+	} );
+
+	const combinations = cartesian( ...variationTypeOptions );
+	return combinations && combinations.map( function( combination ) {
```
I find those `variable && variable.something || defaultValue` statements to be quite confusing. I would rather put `if ( !variationTypeOptions.length ) return [];` a few lines earlier instead.

justinshreve commented:
@DanReyLop Thanks for the feedback. I've made a few changes based on your comments:

* Destructure shortcut for getting attributes from the product object.
* Switched to the [solution here](http://stackoverflow.com/a/29585704) and cleaned it up a bit to fit our coding standards. I'm going to avoid introducing an npm library for this right now but if we find out we need to do more of these operations in other places and can make a case for it, we can easily swap out a solution later.
* Build the name/option combinations first.

Do you mind taking another look?

DanReyLop commented on client/extensions/woocommerce/lib/generate-variations/index.js:
```diff
@@ -1,36 +1,50 @@
 
-	const combinations = cartesian( ...variationTypeOptions );
-	return combinations && combinations.map( function( combination ) {
+	return cartesian( [ ...variationTypes ] ).map( function( combination ) {
```
`cartesian( [ ...variationTypes ] )` is exactly the same as `cartesian( ...variationTypes )`, there's no need to clone the array here.

DanReyLop commented on client/extensions/woocommerce/lib/generate-variations/index.js:
```diff
@@ -1,36 +1,50 @@
-					option,
-				};
-			} ) || [ { name: variationTypeNames[ 0 ], option: combination } ],
+			attributes: Array.isArray( combination ) && [ ...combination ] || [ combination ],
```
`combination` is always an Array, so returning `{ attributes: combination }` is enough.
